### PR TITLE
log: Clean up election in the operation loop

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
Comments, variable renames, and factoring out some code.
Review commit by commit.

This is in preparation for fixing #1150.

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
Note: This is a cosmetic PR. It's convenient to review commit-by-commit.

File: log/operation_manager.go [hunk 1]
```diff
@@ -225,22 +228,14 @@ func (o *OperationManager) masterFor(ctx context.Context, allIDs []int64) ([]int
 	// Synchronize the set of log IDs with those we are tracking mastership for.
 	for _, logID := range allStringIDs {
 		knownLogs.Set(1, logID)
-		if o.electionRunner[logID] != nil {
+		if o.runnerCancels[logID] != nil {
```
Comment (Martin2112) [inline log/operation_manager.go:231]: Not sure whether this or `if _, ok := ` is preferred style, or maybe doesn't matter.

File: log/operation_manager.go [hunk 2]
```diff
@@ -225,22 +228,14 @@ func (o *OperationManager) masterFor(ctx context.Context, allIDs []int64) ([]int
...
-			return nil, fmt.Errorf("failed to create election for %v: %v", logID, err)
+			return nil, err
 		}
-		o.electionRunner[logID] = election.NewRunner(logID, &o.info.ElectionConfig, o.tracker, cancel, el)
-		o.runnerWG.Add(1)
-		go func(r *election.Runner) {
-			defer o.runnerWG.Done()
-			r.Run(innerCtx, o.pendingResignations)
-		}(o.electionRunner[logID])
+		o.runnerCancels[logID] = cancel
```
Comment (Martin2112) [inline log/operation_manager.go:238]: This also happens at line 267?

File: log/operation_manager.go (hunk 1 above)
Comment (pavelkalinnikov) [reply log/operation_manager.go:231]: In this case it is equivalent.

File: log/operation_manager.go (hunk 2 above)
Comment (pavelkalinnikov) [reply log/operation_manager.go:238]: Yes. Removed here.

File: log/operation_manager.go [hunk 3]
```diff
@@ -261,6 +256,26 @@ func (o *OperationManager) masterFor(ctx context.Context, allIDs []int64) ([]int
...
 
+func (o *OperationManager) runElection(ctx context.Context, logID string) (context.CancelFunc, error) {
+	glog.Infof("create master election goroutine for %v", logID)
+	cctx, cancel := context.WithCancel(ctx)
+	e, err := o.info.Registry.ElectionFactory.NewElection(cctx, logID)
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to create election for %v: %v", logID, err)
+	}
+	// TODO(pavelkalinnikov): Passing the cancel function is not needed here.
```
Comment (pavelkalinnikov) [inline log/operation_manager.go:267]: Something went wrong during rebase, I'll fix it.

**Comment by pavelkalinnikov** [comment]
All comments are addressed.

**Review by Martin2112 — APPROVED** [review]

**Comment by pavelkalinnikov** [comment]
Interesting: https://travis-ci.org/github/google/trillian/jobs/755698264
`WARNING: DATA RACE`.
I'm taking a look at it.

**Comment by pavelkalinnikov** [comment]
Got it: `NewRunner` takes `ElectionConfig` by pointer, and tries to modify it in `Run` method. The config is also used in other places. It probably shouldn't be used by pointer.

**Comment by Martin2112** [comment]
Are the modifications in `NewRunner` necessary for it to work? It might not be entirely straightforward to fix.

**Comment by Martin2112** [comment]
OK that sounds like it should work.

**Comment by pavelkalinnikov** [comment]
There is an unrelated test failure `TestLogSuite/DequeueLeavesTwoBatches` for the in-memory CloudSpanner instance. It's flaky because the in-memory CloudSpanner doesn't actually implement transactions.

**Comment by Martin2112** [comment]
OK.

**Comment by pavelkalinnikov** [comment]
Added a commit with a minor comment to trigger builds again.

**Pull request merged**